File: solutions/opus-1/opus/python/brackets.py
import sys

PAIRS = {')': '(', ']': '[', '}': '{'}


def balanced(line):
    stack = []
    for ch in line:
        if ch in '([{':
            stack.append(ch)
        elif ch in PAIRS:
            if not stack or stack.pop() != PAIRS[ch]:
                return False
    return not stack


def main():
    data = sys.stdin.read()
    out = []
    for line in data.split('\n') if data else []:
        out.append(line.strip())
    if data.endswith('\n'):
        out.pop()
    sys.stdout.write(''.join(('yes' if balanced(l) else 'no') + '\n' for l in out))


main()

File: solutions/opus-1/opus/javascript/brackets.js
const data = require('fs').readFileSync(0, 'utf8');
const lines = data.split('\n');
if (lines.length > 1 && lines[lines.length - 1] === '') lines.pop();
const pairs = { ')': '(', ']': '[', '}': '{' };
const out = [];
for (const raw of lines) {
  const line = raw.replace(/\r$/, '');
  const stack = [];
  let ok = true;
  for (const c of line) {
    if (c === '(' || c === '[' || c === '{') stack.push(c);
    else if (c in pairs) {
      if (stack.pop() !== pairs[c]) { ok = false; break; }
    }
  }
  out.push(ok && stack.length === 0 ? 'yes' : 'no');
}
process.stdout.write(out.join('\n') + '\n');
